File: lab7/2.py
import numpy as np
import matplotlib.pyplot as plt


# Функция для вычисления производной dy/dx
def dydx(x, y):
    return -2 * x * y + x * np.exp(-(x**2))


# Функция для вычисления точного решения
def exact_solution(x):
    return (np.exp(-(x**2)) * x**2) / 2


# Метод Эйлера-Коши для численного решения дифференциального уравнения
def euler_cauchy(x0, y0, xn, h):
    x = np.arange(x0, xn + h, h)  # Создание массива значений x
    y = np.zeros_like(x)  # Массив значений y (численное решение)
    y_exact = np.zeros_like(x)  # Массив значений точного решения
    error = np.zeros_like(x)  # Массив значений ошибки

    y[0] = y0  # Начальное значение y
    y_exact[0] = exact_solution(x0)  # Начальное значение точного решения
    error[0] = np.abs(y[0] - y_exact[0])  # Начальная ошибка

    # Основной цикл метода Эйлера-Коши
    for i in range(1, len(x)):
        y_pred = y[i - 1] + h * dydx(x[i - 1], y[i - 1])  # Предсказанное значение y
        y[i] = y[i - 1] + (h / 2) * (
            dydx(x[i - 1], y[i - 1]) + dydx(x[i], y_pred)
        )  # Корректировка y
        y_exact[i] = exact_solution(x[i])  # Вычисление точного решения
        error[i] = np.abs(y[i] - y_exact[i])  # Вычисление ошибки

    return x, y, y_exact, error


# Начальные условия и шаг интегрирования
x0 = 0.0
y0 = 0.0
xn = 1.0
h = 0.1

# Получение численного и точного решений, а также ошибок
x, y, y_exact, error = euler_cauchy(x0, y0, xn, h)

# Вывод результатов
print("x\t\ty (Эйлер-Коши)\ty (Точное)\tРазница")
for i in range(len(x)):
    print(f"{x[i]:.1f}\t\t{y[i]:.6f}\t\t{y_exact[i]:.6f}\t\t{error[i]:.6f}")

# Построение графика численного и точного решений
plt.plot(x, y, "bo-", label="Эйлер-Коши")
plt.plot(x, y_exact, "r--", label="Точное")
plt.xlabel("x")
plt.ylabel("y")
plt.legend()
plt.title("Эйлер-Коши и Точное решение")
plt.grid()
plt.show()

File: lab7/2.cpp
#include <cmath>
#include <iomanip>
#include <iostream>
#include <vector>

// Функция для вычисления правой части дифференциального уравнения
double f(double x, double y) { return -2 * x * y + x * exp(-x * x); }

int main() {
  // Начальные условия
  double x0 = 0.0;
  double y0 = 0.0;
  double h = 0.1;
  double xn = 1.0;

  // Число шагов
  int n = static_cast<int>((xn - x0) / h);

  // Векторы для хранения значений x и y
  std::vector<double> x(n + 1);
  std::vector<double> y(n + 1);
  x[0] = x0;
  y[0] = y0;

  // Реализация метода Эйлера-Коши
  for (int i = 0; i < n; ++i) {
    double k1 = h * f(x[i], y[i]);
    double k2 = h * f(x[i] + h, y[i] + k1);
    y[i + 1] = y[i] + 0.5 * (k1 + k2);
    x[i + 1] = x[i] + h;
  }

  // Вычисление точного решения
  std::vector<double> y_exact(n + 1);
  std::vector<double> errors(n + 1);
  for (int i = 0; i <= n; ++i) {
    y_exact[i] = (exp(-x[i] * x[i]) * x[i] * x[i]) / 2;
    errors[i] = std::abs(y[i] - y_exact[i]);
  }

  // Вывод таблицы
  std::cout << std::setw(10) << "x" << std::setw(15) << "y (Euler-Cauchy)"
            << std::setw(15) << "y (Exact)" << std::setw(15) << "Error"
            << std::endl;
  for (int i = 0; i <= n; ++i) {
    std::cout << std::setw(10) << x[i] << std::setw(15) << y[i] << std::setw(15)
              << y_exact[i] << std::setw(15) << errors[i] << std::endl;
  }

  return 0;
}
